### PR TITLE
 Add type annotations to the certbot package (part 1)

### PR DESCRIPTION
**Pull request by adferrand**
This PR is a part of https://github.com/certbot/certbot/pull/9080

It adds type annotations for all functions in the top level modules of the `certbot` package + the sub-package `certbot.tests`.

### CODE REVIEW

File: certbot/certbot/util.py
Comment (alexzorin) [inline certbot/certbot/util.py:519]: Should this be `Union[str, bytes]`? The function body appears to be written to accept either.

File: certbot/certbot/util.py
Comment (alexzorin) [inline certbot/certbot/util.py:601]: Should `domain` be `Union[str, bytes]`? The function body appears to accept either.

File: certbot/certbot/reverter.py
Comment (alexzorin) [inline certbot/certbot/reverter.py:340]: I guess we did agree to try use generic argument types in the `acme` PR:

```suggestion
    def register_undo_command(self, temporary: bool, command: Iterable[str]) -> None:
```

(+ import)

**Review by alexzorin — CHANGES_REQUESTED** [review]

File: certbot-compatibility-test/certbot_compatibility_test/validator.py
Comment (adferrand) [reply certbot-compatibility-test/certbot_compatibility_test/validator.py:34]: In fact, I tried that and `digest` will try internally to call `encode` on the input... So my cast and inline comment.

File: certbot/certbot/configuration.py
Comment (adferrand) [reply certbot/certbot/configuration.py:68]: In fact `revoke` function in `certbot._internal.main` updates the server value of a `NamespaceConfig` here: https://github.com/certbot/certbot/blob/e8265dbf9c3aad879d2bcae0e95e71ba9facd8ce/certbot/certbot/_internal/main.py#L1223

Mypy does not understand the global delegate of `setattr` done to the embeded `argparse.Namespace` instance defined in a `NamespaceConfig`. So I help it my explicting defining this setter. It has been done already in the past for other mutable fields in `NamespaceConfig`.

File: certbot/certbot/configuration.py
Comment (adferrand) [reply certbot/certbot/configuration.py]: Indeed you are right. I fixed that.

File: certbot/certbot/configuration.py
Comment (adferrand) [reply certbot/certbot/configuration.py:47]: Yes, we can add a type hint. I did that.

File: certbot/certbot/crypto_util.py
Comment (adferrand) [reply certbot/certbot/crypto_util.py]: Indeed. I did that.

File: certbot/certbot/crypto_util.py
Comment (adferrand) [reply certbot/certbot/crypto_util.py]: Fixed.

File: certbot/certbot/interfaces.py
Comment (adferrand) [reply certbot/certbot/interfaces.py]: Fixed.

File: certbot/certbot/main.py
Comment (adferrand) [reply certbot/certbot/main.py:9]: Yes when `main` is called from an CLI entrypoint, it cannot be `None`. But in theory one could call this function directly in some python code, and then it could be `None`.

File: certbot/certbot/util.py
Comment (adferrand) [reply certbot/certbot/util.py:524]: Yes, fixed.

File: certbot/certbot/util.py
Comment (adferrand) [reply certbot/certbot/util.py]: Yes, fixed.

File: certbot/certbot/util.py
Comment (adferrand) [reply certbot/certbot/util.py]: Yes, fixed.

File: certbot/certbot/reverter.py
Comment (adferrand) [reply certbot/certbot/reverter.py]: Totally! By the way, when we need to represent an object that is iterable, sizable and with element access by index, we can use `typing.Sequence` instead of `typing.List` to get more genericity and allow tuples for instance.

However I am not sure to have the time to realign existing types in the module, anyway it would be done in a follow-up PR in my opinion.

File: certbot/certbot/interfaces.py
Comment (adferrand) [reply certbot/certbot/interfaces.py:209]: Yes, a simple list would do the trick. If you do not mind, I would prefer to make these kind of changes in another PR that would be dedicated to realign the documentation (and maybe find a way to automatically check the consistency between sphinx doc and types).

**Comment by adferrand** [comment]
Ok @alexzorin I answered to your review.

File: certbot-compatibility-test/certbot_compatibility_test/validator.py
Comment (alexzorin) [reply certbot-compatibility-test/certbot_compatibility_test/validator.py:34]: OK I see, this is an outstanding issue in pyopenssl. I understand now ^^.

File: certbot/certbot/interfaces.py
Comment (alexzorin) [reply certbot/certbot/interfaces.py:209]: Sure thing.

File: certbot/certbot/ocsp.py
Comment (alexzorin) [inline certbot/certbot/ocsp.py:305]: Should there be a similar assertion for the earlier call to `crypto_util.verify_signed_payload` with `chosen_cert_hash`?

Comment (adferrand) [reply certbot/certbot/ocsp.py:305]: Well, it seems that `mypy` is happy with this one. So it is not necessary I guess.

Comment (alexzorin) [reply certbot/certbot/ocsp.py:305]: OK, fair enough :laughing: .

**Review by alexzorin — APPROVED** [review]
LGTM for this part.

I'm not sure whether the intention is to merge these one by one or wait for all the reviews, so I'm not pressing the button.

**Comment by adferrand** [comment]
I designed these PRs to be mergeable independently. Basically some temporary modifications are required to make this possible, and they should be removed one all the types PRs are merged.

These modifications are only some `cast` or `type: ignore` (so no impact at runtime) and are clearly identified in one single commit in each PR, and with an inline TODO on each modification.

**Comment by alexzorin** [comment]
Sounds good. The only other thing I noticed was the `py.typed` files being added to some (so far) unannotated modules. I assume they are doing something useful as part of this PR but I'm not sure what. If that's the case, happy for you/I to hit the merge button.

**Comment by adferrand** [comment]
Basically the py.typed in other modules is an extra care. If there is no annotation at all, it is basically useless. But if there are, it instructs mypy to take them into account from an installed module.

It is obviously interesting for the distributed modules, but it has in fact an interesting effect during the development, because it is also valid for modules installed during a tox job.

Basically, without it, a `mypy certbot/certbot certbot-apache/certbot-apache` and `mypy certbot/certbot && mypy certbot-apache/certbot-apache` will have a different outcome, because in the former, both codebase are taken into account, while in the latter, mypy rely on the installed modules to check the types of the codebase tested, and without py.typed no type is taken into account.

The addition of the py.typed file is done in all the sub PRs to ensure the type checking is always consistent.

**Pull request merged**
